Last Message Before Death Webhook.

VelonacepsCalyxEggs opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
This PR introduces an optional webhook feature that sends the last in-character (IC) messages of dead players before their death at the end of the round.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
This feature is added mainly for fun, I have taken inspiration from a simillar webhook on a SS13 server. 
It doesn’t have a specific reason, but could add an amusing element to the game by capturing players’ final words. 

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
This is my first PR involving C#. So, if I did anything wrong, I will be glad to learn and improve.

The implementation works as follows:

- When TrySendInGameICMessage is called in ChatSystem.cs, it checks if the player sent the message.
- The message is then added or updated in the dictionary located in LastMessageBeforeDeath.cs.
- At the end of the round, onRoundEnd is called from GameTicker.RoundFlow to send the message using a webhook.

This feature is only active if the webhook link is specified in server_config.toml.

## Media

- [ x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

![изображение](https://github.com/user-attachments/assets/3fcbd34a-f8d9-48a3-8f03-30d0e577d265)
~~I know it says After Death on the image... I am dumb.~~

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
-->
:cl: func_kenobi
- add: Added an optional Webhook that will send messages said by players before their death at the end of the round.

VelonacepsCalyxEggs committed:
Added a new webhook to track player's last IC message.

Signed-off-by: VelonacepsCalyxEggs <ivankuvaev777@gmail.com>

VelonacepsCalyxEggs committed:
Code cleanup.

Signed-off-by: VelonacepsCalyxEggs <ivankuvaev777@gmail.com>

Emisse commented:
very fun

VelonacepsCalyxEggs committed:
Bugfixes and Improvements
- Added conditional compilation for Debugging.
- Webhook is no longer exposed to VV.
- Now using an ordered dict.
- Message is now formatted.
- Time is now trunucated.
- Now checking for maximum size and splitting the message (2000).
- Prevention of a ratelimit if more than 30 messages are sent.

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,103 @@
+            //else
+            //{
+            //    _sawmill.Warning($"Failed to send last messages before death. Status code: {response.StatusCode}");
+            //}
```
Sure! I've added conditional compilation, so if DEBUG is enabled, it will log responses.

VelonacepsCalyxEggs commented on Content.Server/GameTicking/GameTicker.CVars.cs:
```diff
@@ -26,6 +26,11 @@ public sealed partial class GameTicker
         private WebhookIdentifier? _webhookIdentifier;
 
+        [ViewVariables]
+        private string? DiscordLastMessageBeforeDeathWebhook { get; set; }
```
My bad, didn't realize that. Fixed in the latest commit. (I think.)

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,103 @@
+                Content = allMessagesString.ToString()
+            };
+            var discordWebhook = new DiscordWebhook();
+            var response = await discordWebhook.CreateMessage(id, payload);
```
Fair enough!

I've googled some things, and apparently, webhooks can send messages up to 2000 symbols in length.
In case of the messages being more than 2000 symbols, it splits them into multiple parts.

About the ratelimit, Pretty sure 30 messages is the limit for webhooks, if that ever happens, it waits for 60 seconds to send more.

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,103 @@
+                throw new InvalidOperationException("GameTiming is not initialized.");
+            }
+
+            message += $" | at [{_gameTiming.CurTime.ToString()}]";
```
Sure thing! I am now formatting the messages like this:

```C#
var currentTime = _gameTiming.CurTime;
string truncatedTime = $"{currentTime.Hours:D2}:{currentTime.Minutes:D2}:{currentTime.Seconds:D2}";
string formattedMessage = $"[{truncatedTime}] {playerName}: {message}";
```

![изображение](https://github.com/user-attachments/assets/e296860a-6074-4005-9d2a-2ce2329f7f3f)

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,103 @@
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+        private readonly MobStateSystem _mobStateSystem;
+
+        private Dictionary<EntityUid, Dictionary<string, string>> _playerData = new();
```
Makes sense, now using an OrderedDictionary for... order...

VelonacepsCalyxEggs committed:
Update Content.Server/GameTicking/GameTicker.CVars.cs

Remove the leftover [ViewVariables] (I am blind).

Co-authored-by: Chief-Engineer <119664036+Chief-Engineer@users.noreply.github.com>

VelonacepsCalyxEggs committed:
Improvements

- All hardcoded variables are now constants. (well not all, all that needed to be)
- If the message exceeds the MaxICLength, it will be cut off at a random interval.
Example: Someone was giving a speech and got blown to bits.
IC: "I am giving a very interesting 1000
character long spee-"
- Made sure the messages don't get split in half.
- Added a delay between messages.

VelonacepsCalyxEggs committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into deathwebhook

VelonacepsCalyxEggs committed:
Merge branch 'deathwebhook' of https://github.com/VelonacepsCalyxEggs/space-station-14 into deathwebhook

VelonacepsCalyxEggs commented:
I have encountered a problem, IGameTiming.CurTime seems to be the time of how long the server has been up, and not the round time. 
Where can I get the round time from instead?

EmoGarbage404 commented:
> I have encountered a problem, IGameTiming.CurTime seems to be the time of how long the server has been up, and not the round time. Where can I get the round time from instead?

GameTicker has it

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -46,58 +56,87 @@ public async void AddMessage(string playerName, EntityUid player, string message
+                }
+                var payload = new WebhookPayload { Content = message };
+                var response = await discordWebhook.CreateMessage(id, payload);
+                messageCount++;
```
Yeah... 
I've added the delay of 2 seconds between messages. 
I have also decreased the maximum length of the IC message itself, so if it breaches the maximum, it will be randomly cut off below the maximum size.

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -46,58 +56,87 @@ public async void AddMessage(string playerName, EntityUid player, string message
+                messages.Add(message.Substring(i, Math.Min(chunkSize, message.Length - i)));
+            }
+            return messages;
+        }
```
Oh! Missed that, I have now made sure that doesn't happen.

I have skimmed through the convetions, and didn't find anything about this... so... I left it as is for now.

VelonacepsCalyxEggs commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -46,58 +56,87 @@ public async void AddMessage(string playerName, EntityUid player, string message
-            {
-                Content = allMessagesString.ToString()
-            };
+            var messages = SplitMessage(allMessagesString.ToString(), 2000);
```
Done!

VelonacepsCalyxEggs committed:
Use GameTicker instead of IGameTiming.

Signed-off-by: VelonacepsCalyxEggs <ivankuvaev777@gmail.com>

VelonacepsCalyxEggs committed:
Convention check

fixed a place where a constant was not used.

VelonacepsCalyxEggs commented:
I have only one concern as of now. 

What if a player says something inapropriate, and gets kicked / banned from the game? Player's message will still stay in the pool.
I not sure if I should be worried about this or not. 
But if I need to implement this, I would probably go about it like this: 
Remove the EntityUid of the player from the dict when he gets kicked/banned.

But I didn't really look at any admin code, so... not sure how I would go about implementing that without touching some sacred code.

VelonacepsCalyxEggs commented:
> When someone is banned or kicked you could just call a method that finds and removes their message from the message list

Oh, so I suppose there is an event listener for that, that also returns the player's uid when someone is kicked/banned?
I will look into it.

VelonacepsCalyxEggs commented:
> I'm not sure if there's an existing event for those or not, but it should be fairly easy to make an event if it's needed
So... uhh... I've asked around and looked around, and I don't think there is are event listeners for that. 
Should I open a separate PR for this or... just jam it in here?

VelonacepsCalyxEggs commented:
> I'm not sure if there's an existing event for those or not, but it should be fairly easy to make an event if it's needed

So... uhh... I've asked around and looked around, and I don't think there is are event listeners for that. 
Should I open a separate PR for this or... just jam it in here?

VelonacepsCalyxEggs committed:
Remove message on ban.

I had to add an event listener for bans, it seems to work.
But I am not sure I did it properly.

Also some minor fixes, like whisper and emote not registering properly.


Signed-off-by: VelonacepsCalyxEggs <ivankuvaev777@gmail.com>

VelonacepsCalyxEggs committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into deathwebhook

VelonacepsCalyxEggs committed:
Merge branch 'deathwebhook' of https://github.com/VelonacepsCalyxEggs/space-station-14 into deathwebhook

VelonacepsCalyxEggs commented:
> You can just add them with this if they'll be used by this PR

I've added an event listener for bans, so when a person gets banned, the message gets removed from the list.
I have one problem though, I can't find the code for the kick command! I've search far and wide, and the only semi-relevent thing I could find is GhostKick, hell, I even went through each file containing the word "kick" and all I found was some admin interface code. I am confuso.

But, honestly, thinking about it, when someone says something that bad, they'd probably get banned anyways.

metalgearsloth commented on Content.Server/Administration/Managers/BanEventArgs.cs:
```diff
@@ -0,0 +1,39 @@
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Content.Server.Administration.Managers
+{
```
file-scoped namespaces.

metalgearsloth commented on Content.Server/Administration/Managers/BanEventArgs.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace Content.Server.Administration.Managers
+{
+    public class BanEventArgs
+    {
+        public class ServerBanEventArgs : EventArgs
+        {
+            public NetUserId? Target { get; }
+            public string? TargetUsername { get; }
+            public NetUserId? BanningAdmin { get; }
+            public (IPAddress, int)? AddressRange { get; }
+            public ImmutableArray<byte>? Hwid { get; }
+            public uint? Minutes { get; }
+            public NoteSeverity Severity { get; }
+            public string Reason { get; }
+
+            public ServerBanEventArgs(NetUserId? target, string? targetUsername, NetUserId? banningAdmin, (IPAddress, int)? addressRange, ImmutableArray<byte>? hwid, uint? minutes, NoteSeverity severity, string reason)
+            {
+                Target = target;
+                TargetUsername = targetUsername;
+                BanningAdmin = banningAdmin;
+                AddressRange = addressRange;
+                Hwid = hwid;
+                Minutes = minutes;
+                Severity = severity;
+                Reason = reason;
+            }
+        }
+    }
+}
```
This looks hella copy-pasted, you sure you can't just pass the actual ban itself around as this will be cringe to update.

Also sealed classes for new classes it gives compiler warnings if you don't.

VelonacepsCalyxEggs replied:
Indeed, it is... 
Fair enough.

I thought of iterating... but I didn't go with it, since I thought that would need a lot of function calls per player. 
I guess there is a function that check if player is banned by NetUserId or simillar.

VelonacepsCalyxEggs replied:
So... uhh... about iterating. I've found a DB function that checks if a NetUserId/IP Adress/Hardware has an active ban. But the only way (that I found) to get the NetUserId with a EntityUid is through an ActorComponent, which is removed when the player leaves the body... so... I am not sure what to do here...

VelonacepsCalyxEggs replied:
Okay, I thought about this a little bit, and I will probably store the player's NetUserId when it writes a message.
Should be good enough.

metalgearsloth commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,199 @@
+            if (!webhookId.HasValue)
+                return;
+
+            var id = webhookId.Value;
+            StringBuilder allMessagesString = new StringBuilder();
+            foreach (DictionaryEntry player in _playerData)
+            {
+                var playerMessages = player.Value as OrderedDictionary;
+                if (playerMessages != null)
+                {
+                    if (playerMessages["EntityUid"] is EntityUid playerUid && _mobStateSystem.IsDead(playerUid))
+                    {
+                        foreach (DictionaryEntry playerName in playerMessages)
+                        {
+                            if (playerName.Key.ToString() != "EntityUid")
+                            {
+                                allMessagesString.AppendLine($"{playerName.Value}");
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (allMessagesString.ToString().Length == 0)
```
You can just iterate players here and get last messages then check bans without having to even add another event arg.

VelonacepsCalyxEggs replied:
Will do.

metalgearsloth commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,199 @@
+            var messages = SplitMessage(allMessagesString.ToString(), MaxMessageSize);
+            var discordWebhook = new DiscordWebhook();
+            int messageCount = 0;
+            foreach (var message in messages)
+            {
+                if (messageCount >= MaxMessagesPerBatch)
+                {
+                    await Task.Delay(60000); // Wait for 1 minute
+                    messageCount = 0;
+                }
```
I know what Task.Delay does but you didn't leave a comment on why this is even being done.

VelonacepsCalyxEggs replied:
Yeah, my bad, it's there, so if the max messages is reached, we wait for the ratelimit to wear off.
Going to make it clearer in the next commit.

metalgearsloth commented on Content.Server/Chat/LastMessageBeforeDeath.cs:
```diff
@@ -0,0 +1,199 @@
+        }
+
+
+
+        private List<string> SplitMessage(string message, int chunkSize)
+        {
+            var messages = new List<string>();
+            int start = 0;
+
+            while (start < message.Length)
+            {
+                int end = start + chunkSize;
+
+                if (end >= message.Length)
+                {
+                    messages.Add(message.Substring(start));
+                    break;
+                }
+
+                int lastNewLine = message.LastIndexOf('\n', end);
+                if (lastNewLine > start)
+                {
+                    messages.Add(message.Substring(start, lastNewLine - start));
+                    start = lastNewLine + 1;
+                }
+                else
+                {
+                    messages.Add(message.Substring(start, chunkSize));
+                    start += chunkSize;
+                }
+            }
+
+            return messages;
+        }
+
```
This should be on the discord manager I mentioned rather than being a singleton tucked away in here.

metalgearsloth reviewed (changes requested):
I'm gonna be real LastMessageBeforeDeath looks like my cooking rn.

It's somewhat of a symptom of no easy way to do webhooks / discord notifs.

In this case I would just make a generic webhook manager (you don't need to port any existing code to use it) and then make the lastmessagebeforedeath a system which still needs a lot of pruning while still retaining similar functionality.

VelonacepsCalyxEggs commented:
> I'm gonna be real LastMessageBeforeDeath looks like my cooking rn.
> 
> It's somewhat of a symptom of no easy way to do webhooks / discord notifs.
> 
> In this case I would just make a generic webhook manager (you don't need to port any existing code to use it) and then make the lastmessagebeforedeath a system which still needs a lot of pruning while still retaining similar functionality.

I am not sure I quite understand what you mean by "generic webhook manager"?

VelonacepsCalyxEggs committed:
Refactor

Removed:
Reverted addition of the OnServerBanCreated event.

Changes:
LastMessageBeforeDeath is now a system.
Sending messages is now handeled through a LastMessageWebhookManager
located in Discord/Managers/
Added some comments in some places to make code more comprehensive.

Additions:
Added descriptions to all functions added by the PR.

VelonacepsCalyxEggs commented:
@metalgearsloth 
I've refactored the LastMessageBeforeDeath as requested.
Hopefully, I did everything right. It feels and looks right at least.

I've described changes in [749d0cb](https://github.com/space-wizards/space-station-14/pull/30235/commits/749d0cb3a1bd9e7cb827349a70b2a91f0b8cedfc).

VelonacepsCalyxEggs committed:
Code cleanup

Forgot to remove some unused constants.

VelonacepsCalyxEggs commented:
Deleting a (moth sound?) failed a PR check. Peak SS14 coding moment. 
I don't think it's related to any of my changes.

VelonacepsCalyxEggs committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into deathwebhook

VelonacepsCalyxEggs committed:
Merge branch 'space-wizards:master' into deathwebhook

VelonacepsCalyxEggs committed:
Merge branch 'deathwebhook' of https://github.com/VelonacepsCalyxEggs/space-station-14 into deathwebhook

metalgearsloth commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,81 @@
+        /// </summary>
+        /// <param name="webhookId">The identifier of the webhook to send messages to.</param>
+        /// <param name="messages">The list of messages to be sent.</param>
+        /// <returns>A task representing the asynchronous operation.  :nerd:</returns>
+        public async Task SendMessagesAsync(WebhookIdentifier webhookId, List<string> messages)
+        {
+            var discordWebhook = new DiscordWebhook();
+            int messageCount = 0;
```
You shouldn't be constructing the webhook every time, just persist it.

VelonacepsCalyxEggs replied:
Done.

VelonacepsCalyxEggs committed:
Made webhook persistent.
Removed unneeded imports.

PJB3005 commented on Content.Server/GameTicking/GameTicker.CVars.cs:
```diff
@@ -72,6 +74,14 @@ private void InitializeCVars()
                 }
             }, true);
+            //CCVar for DiscordLastMessageBeforeDeathWebhook
+            Subs.CVar(_configurationManager, CCVars.DiscordLastMessageBeforeDeathWebhook, value =>
+            {
+                if (!string.IsNullOrWhiteSpace(value))
+                {
+                    _discord.GetWebhook(value, data => _webhookIdentifierLastMessage = data.ToIdentifier());
+                }
+            }, true);
```
This kind of stuff should be done on `LastMessageBeforeDeathSystem`.

PJB3005 commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    public class LastMessageWebhookManager
+    {
+        private readonly DiscordWebhook _discordWebhook = new DiscordWebhook();
```
You cannot just create a new instance of this manager like this, you need to get it as a dependency.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+                var playerMessages = player.Value as OrderedDictionary;
+                if (playerMessages != null)
+                {
+                    if (playerMessages["EntityUid"] is EntityUid playerUid && _mobStateSystem.IsDead(playerUid))
```
This won't log people that have been gibbed, because `IsDead` will fail for those.

VelonacepsCalyxEggs replied:
Yeah... I am not sure how to go about this... check if an entity still exists and if not consider dead?

PJB3005 replied:
Probably.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+                    {
+                        if (playerMessages["NetUserId"] is NetUserId userNetId)
+                        {
+                            var ban = await _db.GetServerBanAsync(null, userNetId, null);
```
This will take easily up to a minute on some of our servers due to sequential database latency for 80 players. You need to:

1. Probably skip checking this for people who are still connected to the server, as they can't be banned.
2. Do all others in parallel with `Task.WhenAll()`.

VelonacepsCalyxEggs replied:
I feel like the easier option would be to not include messages from disconnected players entirely.
As the system can still be easily be abused by latejoining the game saying whatever and then commiting death. I feel like that's just more things for admins to worry about.
But there are other things that might come from that, like higher retention rates after death, which could lead to more ghost roles being filled in the mean time.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+                allMessagesString.AppendLine($"No messages found.");
+            }
+
+            var messages = _webhookManager.SplitMessage(allMessagesString.ToString(), LastMessageWebhookManager.MaxMessageSize);
```
You are concatenating everything and then splitting it again. This is a mess.

VelonacepsCalyxEggs replied:
My bad, now I add all messages to a list and then format them.

PJB3005 commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,82 @@
+            {
+                if (messageCount >= MaxMessagesPerBatch)
+                {
+                    await Task.Delay(60000); // Wait to avoid rate limiting
```
Make this a constant.

PJB3005 commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,82 @@
+                }
+
+                var payload = new WebhookPayload { Content = message };
+                var response = await _discordWebhook.CreateMessage(webhookId, payload);
```
You are discarding the response here.

VelonacepsCalyxEggs commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+
+        private static readonly Random _random = new Random();
+
+        private object? _userNetId;
```
Uhh... grumblebum. fixed.

VelonacepsCalyxEggs replied:
uhh, confusium, fixed.

VelonacepsCalyxEggs commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+
+            if (message.Length > MaxICLength)
+            {
+                // if the message is bigger than the message length limit, we make it cut off at a random iterval.
+                // Example: Someone was giving a speech and got blown to bits.
```
Agreed, I am now formatting the messages after they all passed needed checks at the end of the round(like having to be connected to a server and being dead).

VelonacepsCalyxEggs commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+            }
+
+            var playerMessages = _playerData[player] as OrderedDictionary ?? throw new InvalidOperationException("Player messages dictionary is not initialized.");
+            playerMessages[playerName] = formattedMessage;
```
I will probably will index by IPlayerSession and then by EntityUID (for each character). I am not sure what conditions might make a person change bodies, but honestly, this brings up more philosophical questions than practical ones.

VelonacepsCalyxEggs committed:
Code refactor (big)

Updated:
LastMessageBeforeDeathSystem is now an EntitySystem
Optimized the addition and formating of messages.
Moved the CCVar for the webhook into the system.

Added:
Checks if the player was gibbed.
Checks if the webhook url is specified.
Added some logging.

Removed:
Removed the Webhook url CCVar from RoundFlow.

VelonacepsCalyxEggs commented:
I have refactored the code as per requests (although not entirely sure).
This turned out to be harder than expected, but I think I managed to get it to work.

Also I think the bot forgot to change the status of the PR after the changes were requested. (it is relevant now I guess.)
Requesting re-review.

VelonacepsCalyxEggs committed:
Fix merge conflict.

VelonacepsCalyxEggs committed:
Fix a foolish mistake...
I mixed up the function's name... fixed now.

PJB3005 commented on Content.Server/Chat/Systems/ChatSystem.cs:
```diff
@@ -78,6 +81,10 @@ public override void Initialize()
         Subs.CVar(_configurationManager, CCVars.LoocEnabled, OnLoocEnabledChanged, true);
         Subs.CVar(_configurationManager, CCVars.DeadLoocEnabled, OnDeadLoocEnabledChanged, true);
         Subs.CVar(_configurationManager, CCVars.CritLoocEnabled, OnCritLoocEnabledChanged, true);
+        Subs.CVar(_configurationManager, CCVars.DiscordLastMessageBeforeDeathWebhook, value =>
+        {
+            if (!string.IsNullOrWhiteSpace(value)) _lastMessageWebhookEnabled = true;
+        }, true);
```
This boolean should be tracked fully in the other system, have `LastMessageBeforeDeathSystem` decide whether to ignore it or not.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,151 @@
+        private static LastMessageBeforeDeathSystem? _instance;
+        private static readonly object _lock = new object();
+
+        private static readonly Random _random = new Random();
```
This is still not fixed.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        private readonly LastMessageWebhookManager _webhookManager = new LastMessageWebhookManager();
+        private Dictionary<ICommonSession, PlayerData> _playerData = new Dictionary<ICommonSession, PlayerData>();
+
+        private const string MaxICLengthCVar = "chat.max_ic_length";
```
1. You should be referring to the `CCVar`.
2. Is what I would say, except this CVar doesn't actually exist.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        [Dependency] private readonly ILogManager _logManager = default!;
+        [Dependency] private readonly DiscordWebhook _discord = default!;
+
+        private MobStateSystem _mobStateSystem = default!;
+        private GameTicker _gameTicker = default!;
```
You can just `[Dependency]` to get these.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+
+        private MobStateSystem _mobStateSystem = default!;
+        private GameTicker _gameTicker = default!;
+        private IEntityManager _entManager = default!;
```
You don't need this, all entity systems already have a reference with their `EntityManager` field.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        private Dictionary<ICommonSession, PlayerData> _playerData = new Dictionary<ICommonSession, PlayerData>();
+
+        private const string MaxICLengthCVar = "chat.max_ic_length";
+        private const int MaxWebhookMessageLength = 2000; // This should not really be changed... since the maximum characters per message for webhooks is 2000.
```
I would still make this a CVar for experimentation/debugging/troubleshooting.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        public override void Initialize()
+        {
+            base.Initialize();
+            IoCManager.InjectDependencies(this);
```
You don't need to do this on entity systems.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        private GameTicker _gameTicker = default!;
+        private IEntityManager _entManager = default!;
+
+        private ISawmill _sawmill = default!;
```
Every entity system already has a sawmill with `Log`, you don't need to request one yourself.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+{
+    internal class LastMessageBeforeDeathSystem : EntitySystem
+    {
+        [Dependency] private readonly IServerDbManager _db = default!;
```
This isn't used.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        private ISawmill _sawmill = default!;
+
+        private readonly LastMessageWebhookManager _webhookManager = new LastMessageWebhookManager();
+        private Dictionary<ICommonSession, PlayerData> _playerData = new Dictionary<ICommonSession, PlayerData>();
```
Don't store this per `ICommonSession`, as reconnecting gives players a new session object. Use `NetUserId` instead.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        private class PlayerData
+        {
+            public Dictionary<string, CharacterData> Characters { get; set; } = new Dictionary<string, CharacterData>();
+            public NetUserId? NetUserId { get; set; }
```
This isn't used.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            public Dictionary<string, CharacterData> Characters { get; set; } = new Dictionary<string, CharacterData>();
+            public NetUserId? NetUserId { get; set; }
+            public ICommonSession? PlayerSession { get; set; }
```
1. You don't need to store this again, you already have the dictionary key available which stores it.
2. Even if the above wasn't important you still have the problem that you shouldn't store these objects.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+
+        private class PlayerData
+        {
+            public Dictionary<string, CharacterData> Characters { get; set; } = new Dictionary<string, CharacterData>();
```
This never needs to be modified, so you can do this.

```suggestion
            public Dictionary<string, CharacterData> Characters { get; } = new Dictionary<string, CharacterData>();
```

PJB3005 replied:
Also you should probably key this on their mind, as that provides the best approximation of a character identity in an IC context.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+
+        private class CharacterData
+        {
+            public string? Message { get; set; } // Store only the last message
```
Rename this to `LastMessage`.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            public string? Message { get; set; } // Store only the last message
+            public EntityUid EntityUid { get; set; }
+            public string? CharacterName { get; set; }
```
Again, you don't need to store this twice.

PJB3005 commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        [Dependency] private readonly DiscordWebhook _discordWebhook = default!;
+
+        public const int MaxMessageSize = 2000; // CAN'T BE MORE THAN 2000! DISCORD LIMIT.
```
This isn't used.

PJB3005 commented on Content.Server/Discord/Managers/LastMessageWebhookManager.cs:
```diff
@@ -0,0 +1,61 @@
+        [Dependency] private readonly DiscordWebhook _discordWebhook = default!;
+
+        public const int MaxMessageSize = 2000; // CAN'T BE MORE THAN 2000! DISCORD LIMIT.
+        private const int MaxMessagesPerBatch = 15;
+        private const int MessageDelayMs = 2000;
+        private const int RateLimitDelayMs = 60000;
```
CVar these.

PJB3005 commented on Content.Server/Chat/Systems/LastMessageBeforeDeathSystem.cs:
```diff
@@ -0,0 +1,189 @@
+        public async void AddMessage(EntityUid source, ICommonSession playerSession, string message, string characterName)
+        {
+
+            if (!_playerData.ContainsKey(playerSession))
+            {
+                _playerData[playerSession] = new PlayerData();
+            }
+
+            var playerData = _playerData[playerSession];
```
You can use `GetOrNew` for this.

PJB3005 commented on Content.Server/Chat/Systems/ChatSystem.cs:
```diff
@@ -743,6 +754,15 @@ private bool CanSendInGame(string message, IConsoleShell? shell = null, ICommonS
+
+        if (player != null && source != null)
+        {
+            _lastMessageBeforeDeathSystem.AddMessage(source, player, message, Identity.Name(source, EntityManager));
```
`Identity.Name` still just reports their perceived name, not their true name. This still isn't correct.

PJB3005 reviewed (changes requested):
Haven't fully reviewed this but well here's enough to keep you busy.

VelonacepsCalyxEggs committed:
Merge remote-tracking branch 'upstream/master' into deathwebhook

# Conflicts:
#	Content.Server/Administration/Managers/IBanManager.cs
#	Content.Shared/CCVar/CCVars.cs

Added a webhook related CCVar in Content.Shared/CCVar/CCVars.Discord.cs

VelonacepsCalyxEggs committed:
Merge branch 'space-wizards:master' into deathwebhook

VelonacepsCalyxEggs committed:
Now keying the user by NetUserId and their character by Mind.
Now using mind to get the Character Name to avoid obfuscation.
Added all constants as CVars.
Webhook CVar is now tracked fully in the LastMessage system.
Removed unneeded junk.

VelonacepsCalyxEggs commented:
I think I did everything specified in the review, but I am not sure about some things, specifically not storing ICommonSession, I need to store it to check whenever the player was banned or kicked during their gameplay, which would disqualify them from getting into the webhook, but, that also could mean that if a player looses connection and reconnects, they would be disqualified too, I need a way to check directly by NetUserId, because it is persistent, I was thinking of checking if there is a mind linked to this NetUserId, but I am not sure what happens to a player's mind when they disconnect.

I need some advice on this.

VelonacepsCalyxEggs committed:
Merge branch 'master' into deathwebhook

AirFryerBuyOneGetOneFree2 commented:
> I think I did everything specified in the review, but I am not sure about some things, specifically not storing ICommonSession, I need to store it to check whenever the player was banned or kicked during their gameplay, which would disqualify them from getting into the webhook, but, that also could mean that if a player looses connection and reconnects, they would be disqualified too, I need a way to check directly by NetUserId, because it is persistent, I was thinking of checking if there is a mind linked to this NetUserId, but I am not sure what happens to a player's mind when they disconnect.
> 
> I need some advice on this.

The new parrot system implemented this!! After I get my impstation port pr merged I may look into upstreaming it. I am currently looking into applying the entity's accent to the last words, as well as capturing the damage values on player death.
